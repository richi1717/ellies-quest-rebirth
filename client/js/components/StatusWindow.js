import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export default function StatusWindow(props) {
  function showStatusPerCharacter() {
    const status = [];
    let incr = 0;

    props.state.characterStats.map((h) => {
      incr++;
      const hpPercentage = Math.ceil((h.currentHp / h.maxHp) * 100);
      const mpPercentage = Math.ceil((h.currentMp / h.maxMp) * 100);
      const lowHealth = {
        'low-health': hpPercentage <= 25
      };
      const lowMagic = {
        'low-magic': mpPercentage <= 25
      };
      status.push(
        <tr key={incr}>
          <td className={`health-bar ${classnames(lowHealth)}`}>
            {h.currentHp}/{h.maxHp}
            <div>
              <span style={{ width: `${hpPercentage}%` }} />
            </div>
          </td>
          <td className={`magic-bar ${classnames(lowMagic)}`}>
            {h.currentMp}/{h.maxMp}
            <div>
              <span style={{ width: `${mpPercentage}%` }} />
            </div>
          </td>
        </tr>
      );
    });
    return status;
  }

  function renderNames() {
    const names = [];
    let incr = 0;

    props.state.characterStats.map((h) => {
      incr++;
      const selected = h.attackerId === props.state.whoIsAttacking.attacker ? 'character-turn' : '';
      names.push(
        <tr key={incr}>
          <td className={`menu-select character ${selected}`}>{h.name}</td>
        </tr>
      );
    });
    return names;
  }

  return (
    <div className="battle-menu-container">
      <div className="battle-menu-main">
        <table>
          <tbody><tr><th>NAME</th></tr></tbody>
          <tbody>{renderNames()}</tbody>
        </table>
      </div>
      <div className="battle-menu-main-stats">
        <table>
          <tbody><tr><th>HP</th><th>MP</th></tr></tbody>
          <tbody>{showStatusPerCharacter()}</tbody>
        </table>
      </div>
      {props.children}
    </div>
  );
}

StatusWindow.propTypes = {
  state: PropTypes.object.isRequired,
  children: PropTypes.node
};
