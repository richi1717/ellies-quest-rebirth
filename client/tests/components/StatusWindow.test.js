import React from 'react';
import { shallow } from 'enzyme';
import StatusWindow from '../../js/components/StatusWindow';
import charactersData from '../testHelpers/charactersData';

describe('<StatusWindow />', () => {
  let statusWindow;

  beforeEach(() => {
    statusWindow = shallow(
      <StatusWindow
        state={{
          characterStats: charactersData,
          whoIsAttacking: {
            attacker: 'hero1'
          }
        }}
      />
    );
  });

  test('should match snapshot', () => {
    expect(statusWindow.getNode()).toMatchSnapshot();
  });

  test('should have a className of battle-menu-container', () => {
    expect(statusWindow).toHaveClassName('battle-menu-container');
  });

  test('should have a className of battle-menu-main', () => {
    expect(statusWindow.find('div').at(1)).toHaveClassName('battle-menu-main');
  });

  test('should have 2 table tags', () => {
    expect(statusWindow.find('table')).toHaveLength(2);
  });

  test('should have a tbody in first table ', () => {
    expect(statusWindow.find('table').at(0).find('tbody'))
      .toHaveLength(2);
  });

  test('should say NAME in first table first th', () => {
    expect(statusWindow.find('th').at(0).text())
      .toEqual('NAME');
  });

  test('should have className of menu-select character in first table second tbody', () => {
    expect(statusWindow.find('table')
      .at(0)
      .find('tbody')
      .at(1)
      .find('td')
      .at(0)).toHaveClassName('menu-select character');
  });

  test('should say HP in second table first th', () => {
    expect(statusWindow
      .find('table')
      .at(1)
      .find('th')
      .at(0)
      .text()).toEqual('HP');
  });

  test('should say MP in second table second th', () => {
    expect(statusWindow.find('table')
      .at(1)
      .find('th')
      .at(1)
      .text()).toEqual('MP');
  });

  test('should have className of health-bar in second table second tbody', () => {
    expect(statusWindow.find('table')
      .at(1)
      .find('tbody')
      .at(1)
      .find('td')
      .at(0)).toHaveClassName('health-bar');
  });

  test('should have className of magic-bar in second table second tbody', () => {
    expect(statusWindow.find('table')
      .at(1)
      .find('tbody')
      .at(1)
      .find('td')
      .at(1)).toHaveClassName('magic-bar');
  });
});

