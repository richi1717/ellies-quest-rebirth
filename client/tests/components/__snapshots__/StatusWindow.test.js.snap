// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatusWindow /> should match snapshot 1`] = `
<div
  className="battle-menu-container"
>
  <div
    className="battle-menu-main"
  >
    <table>
      <tbody>
        <tr>
          <th>
            NAME
          </th>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <td
            className="menu-select character character-turn"
          >
            Link
          </td>
        </tr>
        <tr>
          <td
            className="menu-select character "
          >
            Ellie
          </td>
        </tr>
        <tr>
          <td
            className="menu-select character "
          >
            Luke
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="battle-menu-main-stats"
  >
    <table>
      <tbody>
        <tr>
          <th>
            HP
          </th>
          <th>
            MP
          </th>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <td
            className="health-bar "
          >
            150
            /
            350
            <div>
              <span
                style={
                  Object {
                    "width": "43%",
                  }
                }
              />
            </div>
          </td>
          <td
            className="magic-bar "
          >
            30
            /
            30
            <div>
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
          </td>
        </tr>
        <tr>
          <td
            className="health-bar "
          >
            219
            /
            219
            <div>
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
          </td>
          <td
            className="magic-bar "
          >
            50
            /
            50
            <div>
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
          </td>
        </tr>
        <tr>
          <td
            className="health-bar "
          >
            219
            /
            219
            <div>
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
          </td>
          <td
            className="magic-bar "
          >
            50
            /
            50
            <div>
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
